File: legacy/packages/lsp/src/prompt-preview.ts
import type {
  AgentDecl,
  ConfigItem,
  Expression,
  Program,
} from "@moon/ast";
import { assemblePrompt, type AssembledPrompt } from "@moon/prompt";
import type { JsonSchema } from "@moon/schema-compiler";
import { compileSchemas } from "@moon/schema-compiler";

const ANALYZE_SCHEMA: JsonSchema = {
  type: "object",
  properties: {
    findings: { type: "array", items: { type: "object" } },
    summary: { type: "string" },
    confidence: { type: "number", minimum: 0, maximum: 1 },
  },
  required: ["findings", "summary", "confidence"],
};

export interface PromptSite {
  line: number;
  endLine: number;
  kind: "analyze" | "storm";
  agent: string;
  title: string;
}

export interface PromptPreviewResult {
  site: PromptSite;
  assembled: AssembledPrompt;
  markdown: string;
}

export function findPromptSites(program: Program): PromptSite[] {
  const sites: PromptSite[] = [];

  for (const decl of program.declarations) {
    if (decl.kind !== "Function") continue;
    for (const eq of decl.decl.equations) {
      if (!("statements" in eq.body)) continue;
      for (const stmt of eq.body.statements) {
        if (stmt.kind === "Bind") {
          const call = findAnalyzeCall(stmt.expr);
          if (call) {
            sites.push({
              line: stmt.span.start.line - 1,
              endLine: stmt.span.end.line - 1,
              kind: "analyze",
              agent: call.agent,
              title: `Preview prompt: ${call.agent}`,
            });
          }
        }
        if (stmt.kind === "Storm") {
          const synthesizer = configString(stmt.config, "synthesizer") ?? "Synthesizer";
          sites.push({
            line: stmt.span.start.line - 1,
            endLine: stmt.span.end.line - 1,
            kind: "storm",
            agent: synthesizer,
            title: `Preview storm: ${synthesizer}`,
          });
        }
      }
    }
  }

  return sites;
}

function findAnalyzeCall(expr: Expression): { agent: string; input: unknown } | null {
  if (expr.kind === "App" && expr.func.kind === "FieldAccess" && expr.func.field === "analyze") {
    const agent = exprToName(expr.func.object);
    if (!agent) return null;
    return { agent, input: exprToPlaceholder(expr.arg) };
  }
  if (expr.kind === "FieldAccess" && expr.field === "analyze") {
    const agent = exprToName(expr.object);
    if (!agent) return null;
    return { agent, input: "<input>" };
  }
  return null;
}

function exprToName(expr: Expression): string | null {
  if (expr.kind === "Var") return expr.name;
  if (expr.kind === "Paren") return exprToName(expr.expr);
  return null;
}

function exprToPlaceholder(expr: Expression): unknown {
  if (expr.kind === "Lit") {
    if (expr.value.kind === "String") return expr.value.value;
    if (expr.value.kind === "Int" || expr.value.kind === "Float") return expr.value.value;
    if (expr.value.kind === "Bool") return expr.value.value;
  }
  if (expr.kind === "Var") return `<${expr.name}>`;
  if (expr.kind === "List") {
    return expr.elements.map(exprToPlaceholder);
  }
  if (expr.kind === "Record") {
    const obj: Record<string, unknown> = { _type: expr.name };
    for (const f of expr.fields) obj[f.name] = exprToPlaceholder(f.value);
    return obj;
  }
  return "<expr>";
}

function configValues(config: ConfigItem[]): Record<string, unknown> {
  const out: Record<string, unknown> = {};
  for (const item of config) {
    out[item.key] = exprToPlaceholder(item.value);
  }
  return out;
}

function configString(config: ConfigItem[], key: string): string | null {
  const item = config.find((c) => c.key === key);
  if (!item) return null;
  if (item.value.kind === "Var") return item.value.name;
  if (item.value.kind === "Lit" && item.value.value.kind === "String") {
    return item.value.value.value;
  }
  return null;
}

function collectAgents(program: Program): Map<string, AgentDecl> {
  const agents = new Map<string, AgentDecl>();
  for (const decl of program.declarations) {
    if (decl.kind === "Agent") agents.set(decl.decl.name, decl.decl);
  }
  return agents;
}

function agentModel(agent: AgentDecl): string {
  for (const item of agent.config) {
    if (item.key === "model" && item.value.kind === "Lit" && item.value.value.kind === "String") {
      return item.value.value.value;
    }
  }
  return "deepseek-v4-pro";
}

function agentSystemPrompt(agent: AgentDecl): string | undefined {
  for (const item of agent.config) {
    if (item.key === "systemPrompt" && item.value.kind === "Lit" && item.value.value.kind === "String") {
      return item.value.value.value;
    }
  }
  return undefined;
}

function agentRole(agent: AgentDecl): string | undefined {
  for (const item of agent.config) {
    if (item.key === "role" && item.value.kind === "Lit" && item.value.value.kind === "String") {
      return item.value.value.value;
    }
  }
  return undefined;
}

function agentFocus(agent: AgentDecl): string[] | undefined {
  for (const item of agent.config) {
    if (item.key === "focus" && item.value.kind === "List") {
      const out: string[] = [];
      for (const el of item.value.elements) {
        if (el.kind === "Lit" && el.value.kind === "String") out.push(el.value.value);
      }
      return out.length ? out : undefined;
    }
  }
  return undefined;
}

function schemaForAgent(agent: AgentDecl, schemas: Record<string, JsonSchema>): JsonSchema {
  if (agent.type.kind === "Con" && agent.type.name === "Reviewer") {
    return schemas.ReviewResult ?? ANALYZE_SCHEMA;
  }
  return ANALYZE_SCHEMA;
}

export function buildPromptPreview(
  program: Program,
  line: number,
): PromptPreviewResult | null {
  const sites = findPromptSites(program);
  const site = sites.find((s) => s.line === line);
  if (!site) return null;

  const agents = collectAgents(program);
  const schemas = compileSchemas(program).schemas;

  if (site.kind === "storm") {
    return buildStormPreview(program, site, agents, schemas);
  }

  return buildAnalyzePreview(program, site, agents, schemas);
}

function buildAnalyzePreview(
  program: Program,
  site: PromptSite,
  agents: Map<string, AgentDecl>,
  schemas: Record<string, JsonSchema>,
): PromptPreviewResult | null {
  const agent = agents.get(site.agent);
  if (!agent) return null;

  const bind = findBindAtLine(program, site.line);
  if (!bind) return null;

  const call = findAnalyzeCall(bind.expr);
  if (!call) return null;

  const config = configValues(bind.config);
  const bindFocus = Array.isArray(config.focus) ? config.focus as string[] : undefined;

  const assembled = assemblePrompt({
    agent: site.agent,
    model: agentModel(agent),
    systemPrompt: agentSystemPrompt(agent),
    role: agentRole(agent),
    focus: agentFocus(agent) ?? bindFocus,
    input: call.input,
    config,
    schema: schemaForAgent(agent, schemas),
  });

  return {
    site,
    assembled,
    markdown: formatPreviewMarkdown(site, assembled),
  };
}

function buildStormPreview(
  program: Program,
  site: PromptSite,
  agents: Map<string, AgentDecl>,
  schemas: Record<string, JsonSchema>,
): PromptPreviewResult | null {
  const storm = findStormAtLine(program, site.line);
  if (!storm) return null;

  const config = configValues(storm.config);
  const synthesizerName = configString(storm.config, "synthesizer") ?? site.agent;
  const agent = agents.get(synthesizerName);
  if (!agent) return null;

  const panel = Array.isArray(config.panel) ? config.panel as string[] : [];
  const rounds = typeof config.rounds === "number" ? config.rounds : 1;

  const assembled = assemblePrompt({
    agent: synthesizerName,
    model: agentModel(agent),
    systemPrompt: agentSystemPrompt(agent),
    role: agentRole(agent),
    focus: agentFocus(agent),
    input: exprToPlaceholder(storm.input),
    config,
    schema: schemaForAgent(agent, schemas),
    peerOutputs: panel.map((name) => ({
      agent: String(name),
      summary: "<peer output after panel round>",
    })),
  });

  const stormNote = [
    `**Storm panel:** ${panel.join(", ") || "(none)"}`,
    `**Rounds:** ${rounds}`,
    "",
  ].join("\n");

  return {
    site,
    assembled,
    markdown: stormNote + formatPreviewMarkdown(site, assembled),
  };
}

function findBindAtLine(program: Program, line: number) {
  for (const decl of program.declarations) {
    if (decl.kind !== "Function") continue;
    for (const eq of decl.decl.equations) {
      if (!("statements" in eq.body)) continue;
      for (const stmt of eq.body.statements) {
        if (stmt.kind === "Bind" && stmt.span.start.line - 1 === line) {
          return stmt;
        }
      }
    }
  }
  return null;
}

function findStormAtLine(program: Program, line: number) {
  for (const decl of program.declarations) {
    if (decl.kind !== "Function") continue;
    for (const eq of decl.decl.equations) {
      if (!("statements" in eq.body)) continue;
      for (const stmt of eq.body.statements) {
        if (stmt.kind === "Storm" && stmt.span.start.line - 1 === line) {
          return stmt;
        }
      }
    }
  }
  return null;
}

export function formatPreviewMarkdown(site: PromptSite, assembled: AssembledPrompt): string {
  const parts = [
    `# ${site.title}`,
    "",
    `**Temperature:** ${assembled.temperature}${assembled.maxTokens ? ` · max ${assembled.maxTokens} tokens` : ""}`,
    "",
  ];

  for (const msg of assembled.messages) {
    parts.push(`## ${msg.role.toUpperCase()}`, "", "```", msg.content, "```", "");
  }

  return parts.join("\n");
}

File: Sources/MoonLSP/PromptPreview.swift
import Foundation
import MoonAST
import MoonPrompt
import MoonSchemaCompiler

public struct PromptSite: Sendable, Equatable {
    public var line: Int
    public var endLine: Int
    public var kind: String
    public var agent: String
    public var title: String

    public init(line: Int, endLine: Int, kind: String, agent: String, title: String) {
        self.line = line
        self.endLine = endLine
        self.kind = kind
        self.agent = agent
        self.title = title
    }
}

public struct PromptPreviewResult: Sendable, Equatable {
    public var site: PromptSite
    public var assembled: AssembledPrompt
    public var markdown: String

    public init(site: PromptSite, assembled: AssembledPrompt, markdown: String) {
        self.site = site
        self.assembled = assembled
        self.markdown = markdown
    }
}

public func findPromptSites(_ program: Program) -> [PromptSite] {
    var sites: [PromptSite] = []
    for decl in program.declarations {
        guard case .function(let fn, _) = decl else { continue }
        for eq in fn.equations {
            guard case .doBlock(let block) = eq.body else { continue }
            for stmt in block.statements {
                if case .bind(_, let expr, _, let span) = stmt,
                   let call = findAnalyzeCall(expr) {
                    sites.append(PromptSite(
                        line: span.start.line - 1,
                        endLine: span.end.line - 1,
                        kind: "analyze",
                        agent: call.agent,
                        title: "Preview prompt: \(call.agent)"
                    ))
                }
                if case .storm(_, _, let config, let span) = stmt {
                    let synthesizer = configString(config, key: "synthesizer") ?? "Synthesizer"
                    sites.append(PromptSite(
                        line: span.start.line - 1,
                        endLine: span.end.line - 1,
                        kind: "storm",
                        agent: synthesizer,
                        title: "Preview storm: \(synthesizer)"
                    ))
                }
            }
        }
    }
    return sites
}

public func buildPromptPreview(_ program: Program, line: Int) -> PromptPreviewResult? {
    let sites = findPromptSites(program)
    guard let site = sites.first(where: { $0.line == line }) else { return nil }
    let agents = collectPreviewAgents(program)
    let schemas = compileSchemas(program).schemas

    if site.kind == "storm" {
        return buildStormPreview(program, site: site, agents: agents, schemas: schemas)
    }
    return buildAnalyzePreview(program, site: site, agents: agents, schemas: schemas)
}

private func buildAnalyzePreview(
    _ program: Program,
    site: PromptSite,
    agents: [String: AgentDecl],
    schemas: [String: MoonJsonSchema]
) -> PromptPreviewResult? {
    guard let agent = agents[site.agent],
          let bind = findBindAtLine(program, line: site.line),
          case .bind(_, let expr, let bindConfig, _) = bind,
          let call = findAnalyzeCall(expr) else { return nil }

    let config = configStringMap(bindConfig)
    let bindFocus = config["focus"].map { $0.split(separator: ",").map(String.init) }
    let assembled = assemblePrompt(AssemblyInput(
        agent: site.agent,
        model: previewAgentModel(agent),
        systemPrompt: previewAgentSystemPrompt(agent),
        role: previewAgentRole(agent),
        focus: previewAgentFocus(agent) ?? bindFocus,
        input: call.input,
        config: config
    ))

    return PromptPreviewResult(
        site: site,
        assembled: assembled,
        markdown: formatPreviewMarkdown(site: site, assembled: assembled)
    )
}

private func buildStormPreview(
    _ program: Program,
    site: PromptSite,
    agents: [String: AgentDecl],
    schemas: [String: MoonJsonSchema]
) -> PromptPreviewResult? {
    guard let storm = findStormAtLine(program, line: site.line),
          case .storm(_, let input, let stormConfig, _) = storm else { return nil }
    let config = configStringMap(stormConfig)
    let synthesizerName = configString(stormConfig, key: "synthesizer") ?? site.agent
    guard let agent = agents[synthesizerName] else { return nil }

    let panel = (config["panel"] ?? "").split(separator: ",").map { $0.trimmingCharacters(in: .whitespaces) }
    let assembled = assemblePrompt(AssemblyInput(
        agent: synthesizerName,
        model: previewAgentModel(agent),
        systemPrompt: previewAgentSystemPrompt(agent),
        role: previewAgentRole(agent),
        focus: previewAgentFocus(agent),
        input: exprPlaceholder(input),
        config: config,
        peerOutputs: panel.map { ($0, "<peer output after panel round>") }
    ))

    let stormNote = "**Storm panel:** \(panel.joined(separator: ", "))\n\n"
    return PromptPreviewResult(
        site: site,
        assembled: assembled,
        markdown: stormNote + formatPreviewMarkdown(site: site, assembled: assembled)
    )
}

public func formatPreviewMarkdown(site: PromptSite, assembled: AssembledPrompt) -> String {
    var parts = [
        "# \(site.title)",
        "",
        "**Temperature:** \(assembled.temperature)\(assembled.maxTokens.map { " · max \($0) tokens" } ?? "")",
        "",
    ]
    for msg in assembled.messages {
        parts.append("## \(msg.role.uppercased())")
        parts.append("")
        parts.append("```")
        parts.append(msg.content)
        parts.append("```")
        parts.append("")
    }
    return parts.joined(separator: "\n")
}

private struct AnalyzeCall {
    var agent: String
    var input: String
}

private func findAnalyzeCall(_ expr: MoonAST.Expression) -> AnalyzeCall? {
    switch expr {
    case .app(let fn, let arg, _):
        if case .fieldAccess(let object, let field, _) = fn, field == "analyze",
           let agent = exprName(object) {
            return AnalyzeCall(agent: agent, input: exprPlaceholder(arg))
        }
    case .fieldAccess(let object, let field, _):
        if field == "analyze", let agent = exprName(object) {
            return AnalyzeCall(agent: agent, input: "<input>")
        }
    default:
        break
    }
    return nil
}

private func findBindAtLine(_ program: Program, line: Int) -> DoStatement? {
    for decl in program.declarations {
        guard case .function(let fn, _) = decl else { continue }
        for eq in fn.equations {
            guard case .doBlock(let block) = eq.body else { continue }
            for stmt in block.statements {
                if case .bind = stmt, stmt.span.start.line - 1 == line {
                    return stmt
                }
            }
        }
    }
    return nil
}

private func findStormAtLine(_ program: Program, line: Int) -> DoStatement? {
    for decl in program.declarations {
        guard case .function(let fn, _) = decl else { continue }
        for eq in fn.equations {
            guard case .doBlock(let block) = eq.body else { continue }
            for stmt in block.statements {
                if case .storm = stmt, stmt.span.start.line - 1 == line {
                    return stmt
                }
            }
        }
    }
    return nil
}

private func collectPreviewAgents(_ program: Program) -> [String: AgentDecl] {
    var agents: [String: AgentDecl] = [:]
    for decl in program.declarations {
        if case .agent(let agent, _) = decl {
            agents[agent.name] = agent
        }
    }
    return agents
}

private func exprName(_ expr: MoonAST.Expression) -> String? {
    switch expr {
    case .varRef(let name, _): return name
    case .paren(let inner, _): return exprName(inner)
    default: return nil
    }
}

private func exprPlaceholder(_ expr: MoonAST.Expression) -> String {
    switch expr {
    case .lit(.string(let text, _), _): return text
    case .lit(.int(let n, _), _): return String(n)
    case .lit(.float(let f, _), _): return String(f)
    case .lit(.bool(let b, _), _): return String(b)
    case .varRef(let name, _): return "<\(name)>"
    case .list(let elements, _): return elements.map(exprPlaceholder).joined(separator: ", ")
    case .record(let name, let fields, _):
        let parts = fields.map { "\($0.name)=\(exprPlaceholder($0.value))" }
        return "\(name){\(parts.joined(separator: ", "))}"
    default: return "<expr>"
    }
}

private func configStringMap(_ config: [ConfigItem]) -> [String: String] {
    var out: [String: String] = [:]
    for item in config {
        out[item.key] = exprPlaceholder(item.value)
    }
    return out
}

private func configString(_ config: [ConfigItem], key: String) -> String? {
    guard let item = config.first(where: { $0.key == key }) else { return nil }
    if case .varRef(let name, _) = item.value { return name }
    if case .lit(.string(let text, _), _) = item.value { return text }
    return nil
}

private func previewAgentModel(_ agent: AgentDecl) -> String {
    for item in agent.config where item.key == "model" {
        if case .lit(.string(let model, _), _) = item.value { return model }
    }
    return "deepseek-v4-pro"
}

private func previewAgentSystemPrompt(_ agent: AgentDecl) -> String? {
    for item in agent.config where item.key == "systemPrompt" {
        if case .lit(.string(let text, _), _) = item.value { return text }
    }
    return nil
}

private func previewAgentRole(_ agent: AgentDecl) -> String? {
    for item in agent.config where item.key == "role" {
        if case .lit(.string(let text, _), _) = item.value { return text }
    }
    return nil
}

private func previewAgentFocus(_ agent: AgentDecl) -> [String]? {
    for item in agent.config where item.key == "focus" {
        if case .list(let elements, _) = item.value {
            let values = elements.compactMap { el -> String? in
                if case .lit(.string(let text, _), _) = el { return text }
                return nil
            }
            return values.isEmpty ? nil : values
        }
    }
    return nil
}

private extension DoStatement {
    var span: Span {
        switch self {
        case .bind(_, _, _, let span): return span
        case .storm(_, _, _, let span): return span
        case .letBind(_, _, let span): return span
        case .action(_, _, let span): return span
        }
    }
}